python/cog/server/worker.py

```diff
@@ -171,18 +178,24 @@ def run(self) -> None:
             [ws_stdout, ws_stderr], self._stream_write_hook
         )
         self._stream_redirector.start()
-
         self._setup()
-        asyncio.run(self._loop())
+        self._loop()
         self._stream_redirector.shutdown()
 
+    # I would rather have _setup_async and have a single run_until_complete for setup + loop
+    # but we need to load predictor (and handle errors) to find out the code path
     def _setup(self) -> None:
         done = Done()
         try:
             self._predictor = load_predictor_from_ref(self._predictor_ref)
+            if is_async_predictor(self._predictor):
```

Comment:
The fact that `is_async_predictor` checks both setup and predict makes this a little hard to read IMO.

I think it would be clearer if the setup path did:

```python
self._predictor = load_predictor_from_ref(self._predictor_ref)
# Could be a function or a class
if hasattr(self._predictor, "setup"):
    if is_async(self._predictor.setup):
        get_loop().run_until_complete(run_setup_async(self._predictor))
    else:
        run_setup(self._predictor)
```

and then similarly, later on

```python
def _loop(self) -> None:
        if is_async(get_predict(self._predictor)):
            return get_loop().run_until_complete(self._loop_async())
        return self._loop_sync()
```

Reply:
I think the rationale here is that is predict is async, the user has opted into the async world entirely, and we want to create an event loop before running setup, so get_running_loop in the users' setup code will find the same event loop we're going to use to run _loop_async. that may not be the right approach though

Reply:
Ah, ok, that makes sense. That wasn't clear to me, but the comment you added is very helpful.

Reply:
actually now that I think about it just using new_event_loop doesn't mean that get_running_loop will work, I might be completely wrong...

the crux of the issue is `def setup()` creating a Downloader, which does get_running_loop/new_event_loop, creates a ClientSession implicitly attached to that loop, uses sync_download_files_to_disk, and then (in the future), `async def predict()` might use await `downloader.download_file()` to download a finetune, and ClientSession will error because it's not the same event loop as when it was created. my specific downloader code tries to guard against this but a different user might not do that

maybe if predict is async we should just always wrap setup() in a coroutine so that any get_running_loop will do the right thing?

Reply:
I think what you're doing here should work.

```python
In [24]: async def who_is_running():
    ...:     print(id(asyncio.get_running_loop()))
    ...:

In [25]: await who_is_running()
4338068384

In [26]: loop = asyncio.new_event_loop()

In [27]: id(loop)
Out[27]: 4342403856

In [28]: loop.run_until_complete(who_is_running())
4342403856
```

Reply:
it's more like this

```
def get_loop():
    try:
        return asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.new_event_loop()


async def who_is_running(msg):
    print(msg, id(asyncio.get_running_loop()))


def setup():
    user_loop = get_loop()
    user_loop.run_until_complete(who_is_running("user setup loop"))


async def predict():
    await who_is_running("predict loop")


loop = get_loop()
setup()
loop.run_until_complete(predict())
```

user setup loop 139844438211888
predict loop 139844449272592

if that's IPython with autoawait you can get confusing results

Reply:
We don't have [an existing contract](https://www.hyrumslaw.com/) on `async def setup` functions, so we're free to break people who are spawning their own event loops there. They'll discover and fix that during testing.

We do have an established behavior of allowing event loops in non-async `def setup` or predict functions, disallowing that now will probably break user code.
Maybe that means they should run in another thread so they're free to run their own loop?
(note that we should only call 1 at a time)

Reply:
I think https://github.com/replicate/cog/pull/1366#issuecomment-1811775725 addresses the motivation and the current code in this PR doesn't break people who spawn event loops in setup; running it in a thread shouldn't be needed

Reply:
if neither setup or predict is async, no event loop is created; if predict is async but setup isn't, then an event loop is created but not running, so asyncio.run will work, get_running_loop won't, and people will fall through to new_event_loop. this will be a different event loop than the one that will run an async predict, but that's tolerable